inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/InstrumentationUpdateService.java

```diff
@@ -0,0 +1,176 @@
+        classDisoveryService.addDiscoveryListener(pendingClasses::addAll);
+        // it is possible that we think a class is up to date while it is actually being instrumented with an
+        // outdated instrumentation configuration .for this reason we recheck every class after it has been instrumented.
+        transformer.addPostInstrumentationListener((clazz, conf) -> pendingClasses.add(clazz));
+
+        //all already loaded classes need to be checked if they need instrumentation
+        requestConfigUpdateForAllLoadedClasses();
+
+        startClassCheckerJob();
+
+    }
+
+    private void startClassCheckerJob() {
+        InternalSettings conf = env.getCurrentConfig().getInstrumentation().getInternal();
+        int batchSizes = conf.getClassConfigurationCheckBatchSize();
+        Duration delay = conf.getClassConfigurationCheckInterBatchDelay();
+        classCheckerJob = executor.startJob(this::checkClassesForConfigurationUpdates, batchSizes, delay, delay);
+    }
+
+    /**
+     * Forces all already loaded classes to be checked if their instrumentation is up to date.
+     */
+    private void requestConfigUpdateForAllLoadedClasses() {
+        for (Class<?> clazz : instrumentation.getAllLoadedClasses()) {
+            pendingClasses.add(clazz);
+        }
+    }
+
+    @EventListener
+    private void configEventListener(InspectitConfigChangedEvent ev) {
+        val oldInstrumentation = ev.getOldConfig().getInstrumentation();
+        val newInstrumentation = ev.getNewConfig().getInstrumentation();
+
+        //if the instrumentation configuration has changed all classes need to be reevalauted
+        if (!Objects.equals(oldInstrumentation, newInstrumentation)) {
+            requestConfigUpdateForAllLoadedClasses();
+        }
+
+        InternalSettings newInternal = newInstrumentation.getInternal();
+        classCheckerJob.setBatchSizes(newInternal.getClassConfigurationCheckBatchSize());
+        classCheckerJob.setInterBatchDelay(newInternal.getClassConfigurationCheckInterBatchDelay());
+    }
+
+
+    /**
+     * Takes numberOfClassesToProcess classes from {@link #pendingClasses} and calls {@link #ensureInstrumentationIsUpToDate(Class)}
+     * for them. This method is invoked regularly by {@link #classCheckerJob}.
+     * <p>
+     * This method is package-private for testing.
+     *
+     * @param numberOfClassesToProcess the number of classes to take from {@link #pendingClasses}
+     */
+    void checkClassesForConfigurationUpdates(int numberOfClassesToProcess) {
+        if (transformer.isShuttingDown()) {
+            return; //nothing to do anymore
+        }
+
+        val watch = new StopWatch();
+        try {
+            int processedClassesCount;
+            for (processedClassesCount = 0; processedClassesCount < numberOfClassesToProcess; processedClassesCount++) {
+                Optional<Class<?>> nextClass = CommonUtils.pollElementFromSet(pendingClasses);
```

Comment:
same here, let's poll at least `numberOfClassesToProcess`, maybe then you don't even need the for loop?